### PR TITLE
Create new interrupt-driven i2c driver

### PR DESCRIPTION
**Pull request by Taylorr82**

### CODE REVIEW

File: include/i2c_io.h
Comment (knud) [inline include/i2c_io.h:24]: Curious how these values were chosen? Nothing wrong with them, just wondering if they are related to some other requirements?

File: source/HL_i2c.c
Comment (knud) [inline source/HL_i2c.c:144]: If I understand this correctly, are we sure we want all these interrupts enabled? The NoACK intr was discussed in the Tues advisor meeting, for example, and I can see it, but the Address as slave, Stop Condition, Register Access, and Arbitration Lost interrupts are less obvious to me.

File: source/HL_i2c.c
Comment (knud) [inline source/HL_i2c.c:155]: No need to OR in the value?

Comment (knud) [reply source/HL_i2c.c:155]: If not, perhaps a more descriptive comment...

File: source/HL_i2c.c
Comment (knud) [reply source/HL_i2c.c:144]: Oh, and here and below, perhaps I am commenting on generated code, so the comments may not be totally appropriate...

File: source/HL_i2c.c
Comment (knud) [inline source/HL_i2c.c:223]: See previous comment on enabling all interrupts.

File: source/i2c_io.c
Comment (knud) [inline source/i2c_io.c:37]: In some circles the shortcut IF is frowned on... Not a big fan myself, so just saying.

File: source/i2c_io.c
Comment (knud) [inline source/i2c_io.c:110]: Depending on the answer to enabling this and other interrupts, this may not be needed...

**Review by knud — CHANGES_REQUESTED** [review]
Mostly question about enabling all the interrupts. The there big question is how to stimulate error conditions and thus code coverage. Happy to approve once you let me now about the interrupts question.

File: include/i2c_io.h
Comment (Taylorr82) [reply include/i2c_io.h:24]: More or less at random. I think it's too big but we do have an i2c transfer that requests 100 bytes in another submodule. I don't want the software to timeout when things are actually going fine, just taking a long time

File: source/HL_i2c.c
Comment (Taylorr82) [reply source/HL_i2c.c:144]: I think I could disable the address as slave interrupt. But then I don't know what the behavior will be if the flag gets set in error. My IRQ code explicitly unsets it just in case

Comment (Taylorr82) [reply source/HL_i2c.c:144]: Stop condition is used in every case that a transfer is complete, whether in error or nominally. So it is necessary to give the semaphore letting the calling thread know it is done

Arbitration lost is set so if there's an error where the bus is held low there will be a signal to attempt to recover

Register access seems to be only fired when there's a NACK in the start byte. I think it's supposed to be fired every time a register is ready, as in the TX/RX interrupt for example, but is a lower priority so it doesn't actually fire, then the routines for the other interrupts unset this bit as part of their regular actions. The technical reference for the module shows the ways this interrupt can be cleared

File: source/HL_i2c.c
Comment (Taylorr82) [reply source/HL_i2c.c:155]: This is auto-generated. Nothing I can do unfortunately

File: source/HL_i2c.c
Comment (Taylorr82) [reply source/HL_i2c.c:223]: See previous comment on justification

File: source/i2c_io.c
Comment (Taylorr82) [reply source/i2c_io.c:37]: In this case is the statement simple enough to be acceptable?

**Comment by Taylorr82** [comment]
Testing with a device unplugged was successful. Other types of test are very difficult to conceptualize. A test plan would be ideal (as well as people using i2c devices keeping me updated on any errors then encounter)

File: include/i2c_io.h
Comment (knud) [reply include/i2c_io.h:24]: Well, normally it should never happen and if it does, it probably indicates other issues.

File: source/HL_i2c.c
Comment (knud) [reply source/HL_i2c.c:144]: So, to summarize, if it's not broke...

That said, Register Access seems redundant the way you describe it, so maybe worth losing? Without digging into the data sheet myself in detail, I tend to not enable things just because they are there. The other stuff seems okay.

Your choice on this, just be sure you are comfortable.

File: source/i2c_io.c
Comment (knud) [reply source/i2c_io.c:37]: In this (simple) case it becomes just a question of preference and style. I just wanted to point out my preference. If the logic was more complex, I'd ask to not use it.

**Comment by knud** [comment]
By "A test plan would be ideal" do you mean that a component-level test should be implemented so that multiple I2C clients (aka slaves) can be made to be active and, thus, at least generate a stress test environment? If so, agreed and someone should look into this. Of course, the scenario should be at least the expected operating conditions (e.g., if two peripherals are never enabled at the same time, then no sense trying a test where they are)

Please make note of this and we can close the review.

**Comment by Taylorr82** [comment]
Noted. I think in a more hardware integration test environment this should be stress tested. It does work with a single device on the bus. We will have multiple devices active on each i2c wire at a time, meaning multiple peripherals will be used at once.

File: source/i2c_io.c
Comment (arrooney) [inline source/i2c_io.c:56]: Just have this start as 0

File: include/i2c_io.h
Comment (arrooney) [inline include/i2c_io.h:30]: Maybe we can use a more idiomatic case convention, either i2cSend or i2c_send

File: source/i2c_io.c
Comment (Taylorr82) [inline source/i2c_io.c:83]: Verify proper working of this line

File: include/i2c_io.h
Comment (Taylorr82) [reply include/i2c_io.h:30]: i2cSend won't work because that's the definition of the halcogen drivers. i2c_send is redefined in ADCS. Gonna leave this for now and make it an issue after merging

File: source/i2c_io.c
Comment (joshdellaz) [inline source/i2c_io.c:66]: Would it be worth having more detail in function descriptions? Namely, what the function does and what the arguments are.

Comment (Taylorr82) [reply source/i2c_io.c:66]: Can do

**Pull request merged**
